src/PowerShellEditorServices/Services/DebugAdapter/BreakpointSyncService.cs

```diff
@@ -0,0 +1,904 @@
+        }
+    }
+
+    private record BreakpointTranslationInfo(
+        SyncedBreakpointKind Kind,
+        ScriptBlock? Action = null,
+        string? Name = null,
+        VariableAccessMode VariableMode = default,
+        string? FilePath = null,
+        int Line = 0,
+        int Column = 0);
+
+    internal record struct BreakpointHandle(SemaphoreSlim Target) : IDisposable
+    {
+        public void Dispose() => Target.Release();
+    }
+
+    private readonly BreakpointMap _map = new(new(), new());
+
+    private readonly ConditionalWeakTable<Runspace, BreakpointMap> _attachedRunspaceMap = new();
+
+    private readonly ConcurrentBag<SyncedBreakpoint> _pendingAdds = new();
+
+    private readonly ConcurrentBag<SyncedBreakpoint> _pendingRemovals = new();
+
+    private readonly SemaphoreSlim _breakpointMutateHandle = AsyncUtils.CreateSimpleLockingSemaphore();
+
+    private readonly IInternalPowerShellExecutionService _host;
+
+    private readonly ILogger<BreakpointSyncService> _logger;
+
+    private readonly DebugStateService _debugStateService;
+
+    private readonly ILanguageServerFacade _languageServer;
+
+    private bool? _isSupported;
+
+    public BreakpointSyncService(
+        IInternalPowerShellExecutionService host,
+        ILoggerFactory loggerFactory,
+        DebugStateService debugStateService,
+        ILanguageServerFacade languageServer)
+    {
+        _host = host;
+        _logger = loggerFactory.CreateLogger<BreakpointSyncService>();
+        _debugStateService = debugStateService;
+        _languageServer = languageServer;
+    }
+
+    public bool IsSupported
+    {
+        get => _isSupported ?? false;
+        set
+        {
+            Debug.Assert(_isSupported is null, "BreakpointSyncService.IsSupported should only be set once by initialize.");
+            _isSupported = value;
+        }
+    }
+
+    public bool IsMutatingBreakpoints
+    {
+        get
+        {
+            if (_breakpointMutateHandle.Wait(0))
+            {
+                _breakpointMutateHandle.Release();
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    private int? TargetRunspaceId => _debugStateService?.RunspaceId;
+
+    public async Task UpdatedByServerAsync(BreakpointUpdatedEventArgs eventArgs)
+    {
+        if (!_map.ByPwshId.TryGetValue(eventArgs.Breakpoint.Id, out SyncedBreakpoint existing))
+        {
+            // If we haven't told the client about the breakpoint yet then we can just ignore.
+            if (eventArgs.UpdateType is BreakpointUpdateType.Removed)
+            {
+                return;
+            }
+
+            existing = CreateFromServerBreakpoint(eventArgs.Breakpoint);
+            string? id = await SendToClientAsync(existing.Client, BreakpointUpdateType.Set)
+                .ConfigureAwait(false);
+
+            existing.Client.Id = id!;
+            RegisterBreakpoint(existing);
+            return;
+        }
```

Comment:
Just a little confusing to have this in an `if (!getThing(named out param))` as the out param is _not_ used (since it's not, it wasn't in the map) but we then assign to it locally and use that. I'd do `out Thing _` and then locally declare `Thing breakpoint`. Or did I read this wrong?

Reply:
Yeah it's used outside of that `if` block. The if block is the exit early condition for it not already existing.

I can for sure see how that's confusing though I'll think of a better way to write this

Reply:
Alright I rewrote this to call a nested method and added a new variable instead of reusing `existing` since that was really confusing. LMK what you think!